Update guides for Cy 13, where assertions are now commands (#5081)
cypress-io/cypress-documentation: docs/guides/core-concepts/introduction-to-cypress.mdx

:::info

While Cypress does have a [`.then()`](/api/commands/then) command, Cypress

> commands are not promises and cannot be `await`ed. If you'd like to learn more

Reviewer: Can you update the first mention of the `.then()` command with the command link as well?
```suggestion
commands are not Promises and cannot be `await`ed. If you'd like to learn more
```

:::info

While Cypress is built using Promises that come from

> [Bluebird](http://bluebirdjs.com/), these are not what we expose as commands on

Reviewer 2: This was there before, but it almost feels unnecessary to say "While Cypress is built using Promises that come from
[Bluebird](http://bluebirdjs.com/), these are not what we expose as commands on". We could keep it simple and just say "Commands on `cy` do not return Promises, although they are asynchronous".

Doesn't necessarily need to be updated here, tho

#### You cannot add a `.catch` error handler to a failed command

In Cypress there is no built in error recovery from a failed command. A command
_eventually_ passes, or if it fails, all remaining commands are not executed,
and the test as a whole fails.

You might be wondering:

> How do I create conditional control flow, using if/else? So that if an element
> does (or doesn't) exist, I choose what to do?

Cypress does not support this type of conditional control flow because it leads
to non-deterministic tests - different runs may behave differently, which makes
them less consistent and useful for verifying your application's correctness. In

> general, there are only a handful of very specific situations where you can

Reviewer: ```suggestion
general, there are only a handful of very specific situations where you can or should
```

create control flow using Cypress commands.

> With that said, as long as you are aware of the potential pitfalls with control

Reviewer: lol not sure I know what these pitfalls are... i know you didn't write this, but do you know if there is more info on the conditional testing page about these?

Author: I think they're laid out here - https://docs.cypress.io/guides/core-concepts/conditional-testing#The-problem.

I don't find that page terribly compelling, but not something I want to mess with too much here.

:::

What makes Cypress unique from other testing tools is that assertions
**automatically retry**. Think of them as **guards** - assertions describe what
your application should look like, and Cypress will automatically **block, wait,

> and retry** until it reaches that state.

Reviewer: Should this mention re-executing query commands leading up to the assertion?

Author: Adding some extra words to that effect a couple lines lower in this section.

:::

Without a single explicit assertion, there are dozens of ways this test can
fail. Here's a few:

> - The initial [`cy.mount()`](/api/commands/mount) or

Reviewer: this can't be happening because `.contains()` would fail if mount/visit failed.

<strong>Core Concept</strong>

With Cypress, you don't have to write explicit assertions to have a useful test.
Without a single `expect()` or `.should()`, a few lines of Cypress can ensure

> thousands of lines of code are working properly across the client and server.

Reviewer: I actually like the explication points in this guide since it's the intro and trying to get you hyped 🙃

Author: Hm, I tend to prefer a more steady tone - hype turns me away from docs, and would prefer for features to speak for themselves.

This is because many commands have a built in

> [Implicit Assertions](#Implicit-Assertions) which offer you a high level of

Reviewer: feels like overkill to link to the next section.

Author: True. Removing link.

confidence that your application is working as expected.

:::

### Implicit Assertions

Many commands have default, built-in assertions, or rather have requirements

> that may cause it to fail without needing an explicit assertion you've added.

Reviewer: should we mention each command will outline these in their docs?

Author: I'm not sure that needs to be said explicitly - just feels like words for the sake of words.

Would you as a new user *not* expect each command page's to explain it?

> With Cypress, you don't have to assert to have a useful test. Even without

Reviewer 2: I feel like this sentence could maybe use a couple qualifiers, though I do see you're trying to make things less wordy:

```suggestion
With Cypress, you don't always have to explicitly assert to have a useful test. Even without
```

Again, it wasn't written in this PR, so feel free to ignore.

```js

> cy

Reviewer: above where it says "all dom commands automatically wait" should that be queries?

Author: No; there are commands that interact with the DOM (like `.click()`) that aren't queries which wait, and queries (like `.readFile()`) which aren't related to the DOM at all.

  .click()
```

Cypress will automatically _wait_ for elements to pass their implicit
assertions. Like with the explicit assertions you've added, all of these

> assertions share the _same_ timeout values.

Reviewer 2: Maybe it'd be nice to mention where these timeout values come from.

Author: Rewrote these sentences.

> Cypress will automatically _wait_ for elements to pass their implicit assertions. See [Timeouts](#Timeouts) below for more on how timeouts are determined.

<strong>Core Concept</strong>

> If you add [`.should('not.exist')`](/api/commands/should) to any DOM command,
> Cypress will skip its default existence assertion.

Reviewer 2: I had to re-read this to understand it. Maybe it could be made more clear?

```suggestion
If you add [`.should('not.exist')`](/api/commands/should) to any DOM command,
Cypress will skip the default assertion, which is to check existence.
```

Not even sure that's grammatically correct... I asked ChatGPT:

![image](https://user-images.githubusercontent.com/1151760/220998750-db4e64d3-a08f-4501-86a3-a1c3de43bb52.png)

I kinda like the last one:

```suggestion
If you want to disable the default existence assertion for an element in Cypress, you can add [`.should('not.exist')`](/api/commands/should) to any DOM command.
```

Author: Went with
> If you want to disable the default existence assertion, you can add [`.should('not.exist')`](/api/commands/should) to any DOM command.

commands - the `timeout` modification goes on the previous commands _not the
assertions_.

> #### Example #1: Implicit Assertion

Reviewer: Above the section:
```
### Writing Assertions
There are two ways to write assertions in Cypress:
1. **Implicit Subjects:** Using [`.should()`](/api/commands/should) or
   [`.and()`](/api/commands/and).
2. **Explicit Subjects:** Using `expect`.
```

why is .should()/.and() an implicit subject?  it'd just be a queued assertion vs sync assertion right?

Author: Yeah, I think that section could use a touch-up. Rewriting it.

Reviewer 3: since https://github.com/cypress-io/cypress/pull/25738 is no longer making its way into the v13 release, I think this suggestion is no longer applicable.
